src/equal.c

```diff
@@ -19,7 +19,7 @@ int equal_scalar(SEXP x, R_len_t i, SEXP y, R_len_t j, bool na_equal) {
   default: break;
   }
 
-  switch (vec_typeof(x)) {
+  switch (vec_proxy_typeof(x)) {
   case vctrs_type_list: return list_equal_scalar(x, i, y, j, na_equal);
   case vctrs_type_dataframe: return df_equal_scalar(x, i, y, j, na_equal);
```

Comment:
Do we also need to think about tibbles here?

Reply:
The proxy type of a tibble is `vctrs_type_dataframe`.